server/src/test/java/org/elasticsearch/common/geo/EdgeTreeTests.java

```diff
@@ -0,0 +1,135 @@
+            // box-query touches bottom-right corner
+            assertTrue(reader.containedInOrCrosses(maxX, minY - randomIntBetween(1, 180), maxX + randomIntBetween(1, 180), minY));
+            // box-query touches top-right corner
+            assertTrue(reader.containedInOrCrosses(maxX, maxY, maxX + randomIntBetween(1, 180), maxY + randomIntBetween(1, 180)));
+            // box-query touches top-left corner
+            assertTrue(reader.containedInOrCrosses(minX - randomIntBetween(1, 180), maxY, minX, maxY + randomIntBetween(1, 180)));
+            // box-query fully-enclosed inside rectangle
+            assertTrue(reader.containedInOrCrosses((3 * minX + maxX) / 4, (3 * minY + maxY) / 4, (3 * maxX + minX) / 4,
+                (3 * maxY + minY) / 4));
+            // box-query fully-contains poly
+            assertTrue(reader.containedInOrCrosses(minX - randomIntBetween(1, 180), minY - randomIntBetween(1, 180),
+                maxX + randomIntBetween(1, 180), maxY + randomIntBetween(1, 180)));
+            // box-query half-in-half-out-right
+            assertTrue(reader.containedInOrCrosses((3 * minX + maxX) / 4, (3 * minY + maxY) / 4, maxX + randomIntBetween(1, 1000),
+                (3 * maxY + minY) / 4));
+            // box-query half-in-half-out-left
+            assertTrue(reader.containedInOrCrosses(minX - randomIntBetween(1, 1000), (3 * minY + maxY) / 4, (3 * maxX + minX) / 4,
+                (3 * maxY + minY) / 4));
+            // box-query half-in-half-out-top
+            assertTrue(reader.containedInOrCrosses((3 * minX + maxX) / 4, (3 * minY + maxY) / 4, maxX + randomIntBetween(1, 1000),
+                maxY + randomIntBetween(1, 1000)));
+            // box-query half-in-half-out-bottom
+            assertTrue(reader.containedInOrCrosses((3 * minX + maxX) / 4, minY - randomIntBetween(1, 1000),
+                maxX + randomIntBetween(1, 1000), (3 * maxY + minY) / 4));
+
+            // box-query outside to the right
+            assertFalse(reader.containedInOrCrosses(maxX + randomIntBetween(1, 1000), minY, maxX + randomIntBetween(1001, 2000), maxY));
+            // box-query outside to the left
+            assertFalse(reader.containedInOrCrosses(maxX - randomIntBetween(1001, 2000), minY, minX - randomIntBetween(1, 1000), maxY));
+            // box-query outside to the top
+            assertFalse(reader.containedInOrCrosses(minX, maxY + randomIntBetween(1, 1000), maxX, maxY + randomIntBetween(1001, 2000)));
+            // box-query outside to the bottom
+            assertFalse(reader.containedInOrCrosses(minX, minY - randomIntBetween(1001, 2000), maxX, minY - randomIntBetween(1, 1000)));
+        }
+    }
+
+    public void testSimplePolygon() throws IOException  {
+        for (int iter = 0; iter < 1000; iter++) {
+            ShapeBuilder builder = RandomShapeGenerator.createShape(random(), RandomShapeGenerator.ShapeType.POLYGON);
```

Comment:
Since we're just creating Lucene `Polygon`, I think we could also use Lucene's `GeoTestUtil` here so we don't have to rely on the builder. `GeoTestUtil.nextPolygon()` will create a lot of adversarial cases for us to stress test the edge tree.